File: build-tools/minify.py
#!/usr/bin/env python3

import os
import re
import json
from pathlib import Path

def minify_css(content):
    """Simple CSS minification"""
    # Remove comments
    content = re.sub(r'/\*.*?\*/', '', content, flags=re.DOTALL)
    # Remove unnecessary whitespace
    content = re.sub(r'\s+', ' ', content)
    # Remove whitespace around specific characters
    content = re.sub(r'\s*([{}:;,>+~])\s*', r'\1', content)
    # Remove trailing semicolons before closing braces
    content = re.sub(r';}', '}', content)
    return content.strip()

def minify_js(content):
    """Simple JavaScript minification"""
    # Remove single-line comments (preserve URLs)
    content = re.sub(r'//(?![^\n]*https?:)[^\n]*', '', content)
    # Remove multi-line comments
    content = re.sub(r'/\*.*?\*/', '', content, flags=re.DOTALL)
    # Remove unnecessary whitespace
    content = re.sub(r'\s+', ' ', content)
    return content.strip()

def main():
    print('🔧 Starting minification process...')
    
    # Create output directory
    dist_dir = Path('assets/dist')
    dist_dir.mkdir(exist_ok=True)
    
    # Process CSS files
    print('🎨 Minifying CSS files...')
    css_files = []
    css_dir = Path('assets/css')
    
    for css_file in css_dir.glob('*.css'):
        with open(css_file, 'r', encoding='utf-8') as f:
            content = f.read()
        
        minified = minify_css(content)
        output_file = dist_dir / f'{css_file.stem}.min.css'
        
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(minified)
        
        css_files.append(str(output_file))
        savings = ((len(content) - len(minified)) / len(content) * 100) if content else 0
        print(f'  ✓ {css_file.name}: {len(content)} → {len(minified)} bytes ({savings:.1f}% reduction)')
    
    # Process JS files
    print('📜 Minifying JavaScript files...')
    js_files = []
    js_dir = Path('assets/js')
    
    for js_file in js_dir.glob('*.js'):
        with open(js_file, 'r', encoding='utf-8') as f:
            content = f.read()
        
        minified = minify_js(content)
        output_file = dist_dir / f'{js_file.stem}.min.js'
        
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write(minified)
        
        js_files.append(str(output_file))
        savings = ((len(content) - len(minified)) / len(content) * 100) if content else 0
        print(f'  ✓ {js_file.name}: {len(content)} → {len(minified)} bytes ({savings:.1f}% reduction)')
    
    # Create combined files
    print('📦 Creating combined files...')
    
    # Combined CSS
    css_order = ['main.css', 'components.css', 'responsive.css']
    combined_css = []
    
    for css_name in css_order:
        css_path = dist_dir / f'{Path(css_name).stem}.min.css'
        if css_path.exists():
            with open(css_path, 'r', encoding='utf-8') as f:
                combined_css.append(f'/* {css_name} */\n{f.read()}')
    
    combined_css_content = '\n\n'.join(combined_css)
    combined_css_path = dist_dir / 'combined.min.css'
    
    with open(combined_css_path, 'w', encoding='utf-8') as f:
        f.write(combined_css_content)
    
    print(f'  ✓ Combined CSS: {len(combined_css_content)} bytes')
    
    # Combined JS
    js_order = ['image-optimizer.js', 'main.js', 'projects.js', 'publications.js', 
                'technical-artifacts.js', 'about.js', 'contact.js', 'performance.js']
    combined_js = []
    
    for js_name in js_order:
        js_path = dist_dir / f'{Path(js_name).stem}.min.js'
        if js_path.exists():
            with open(js_path, 'r', encoding='utf-8') as f:
                combined_js.append(f'/* {js_name} */\n{f.read()}')
    
    combined_js_content = '\n\n'.join(combined_js)
    combined_js_path = dist_dir / 'combined.min.js'
    
    with open(combined_js_path, 'w', encoding='utf-8') as f:
        f.write(combined_js_content)
    
    print(f'  ✓ Combined JS: {len(combined_js_content)} bytes')
    
    # Create manifest
    manifest = {
        'generated': '2024-12-19T12:00:00.000Z',
        'version': '1.0.0',
        'files': {
            'css': {
                'combined': 'assets/dist/combined.min.css',
                'individual': css_files
            },
            'js': {
                'combined': 'assets/dist/combined.min.js',
                'individual': js_files
            }
        }
    }
    
    manifest_path = dist_dir / 'build-manifest.json'
    with open(manifest_path, 'w', encoding='utf-8') as f:
        json.dump(manifest, f, indent=2)
    
    print('✓ Build manifest created')
    print('✅ Minification completed successfully!')

if __name__ == '__main__':
    main()

File: build-tools/minify.js
#!/usr/bin/env node

/**
 * CSS and JavaScript Minification Tool
 * Optimized for technical content display
 */

const fs = require('fs').promises;
const path = require('path');

class MinificationTool {
    constructor() {
        this.cssDir = './assets/css';
        this.jsDir = './assets/js';
        this.outputDir = './assets/dist';
    }

    async init() {
        console.log('🔧 Starting minification process...');
        
        try {
            await this.ensureOutputDirectory();
            await this.minifyCSS();
            await this.minifyJS();
            await this.generateManifest();
            
            console.log('✅ Minification completed successfully!');
        } catch (error) {
            console.error('❌ Minification failed:', error);
            process.exit(1);
        }
    }

    async ensureOutputDirectory() {
        try {
            await fs.access(this.outputDir);
        } catch {
            await fs.mkdir(this.outputDir, { recursive: true });
            console.log(`📁 Created output directory: ${this.outputDir}`);
        }
    }

    async minifyCSS() {
        console.log('🎨 Minifying CSS files...');
        
        const cssFiles = await this.findFiles(this.cssDir, '.css');
        const minifiedFiles = [];
        
        for (const cssFile of cssFiles) {
            const minified = await this.minifyCSSFile(cssFile);
            minifiedFiles.push(minified);
        }
        
        // Create combined CSS file
        await this.createCombinedCSS(minifiedFiles);
        
        console.log(`✓ Minified ${cssFiles.length} CSS files`);
    }

    async minifyJS() {
        console.log('📜 Minifying JavaScript files...');
        
        const jsFiles = await this.findFiles(this.jsDir, '.js');
        const minifiedFiles = [];
        
        for (const jsFile of jsFiles) {
            const minified = await this.minifyJSFile(jsFile);
            minifiedFiles.push(minified);
        }
        
        // Create combined JS file
        await this.createCombinedJS(minifiedFiles);
        
        console.log(`✓ Minified ${jsFiles.length} JavaScript files`);
    }

    async findFiles(dir, extension) {
        const files = [];
        
        async function scanDirectory(currentDir) {
            const entries = await fs.readdir(currentDir, { withFileTypes: true });
            
            for (const entry of entries) {
                const fullPath = path.join(currentDir, entry.name);
                
                if (entry.isDirectory()) {
                    await scanDirectory(fullPath);
                } else if (entry.isFile() && entry.name.endsWith(extension)) {
                    files.push(fullPath);
                }
            }
        }
        
        await scanDirectory(dir);
        return files;
    }

    async minifyCSSFile(filePath) {
        const content = await fs.readFile(filePath, 'utf8');
        const fileName = path.basename(filePath, '.css');
        
        // Basic CSS minification
        const minified = content
            // Remove comments
            .replace(/\/\*[\s\S]*?\*\//g, '')
            // Remove unnecessary whitespace
            .replace(/\s+/g, ' ')
            // Remove whitespace around specific characters
            .replace(/\s*([{}:;,>+~])\s*/g, '$1')
            // Remove trailing semicolons before closing braces
            .replace(/;}/g, '}')
            // Remove empty rules
            .replace(/[^{}]+{\s*}/g, '')
            .trim();
        
        const outputPath = path.join(this.outputDir, `${fileName}.min.css`);
        await fs.writeFile(outputPath, minified, 'utf8');
        
        const originalSize = Buffer.byteLength(content, 'utf8');
        const minifiedSize = Buffer.byteLength(minified, 'utf8');
        const savings = ((originalSize - minifiedSize) / originalSize * 100).toFixed(1);
        
        console.log(`  ✓ ${fileName}.css: ${originalSize} → ${minifiedSize} bytes (${savings}% reduction)`);
        
        return {
            original: filePath,
            minified: outputPath,
            content: minified,
            originalSize,
            minifiedSize,
            savings
        };
    }

    async minifyJSFile(filePath) {
        const content = await fs.readFile(filePath, 'utf8');
        const fileName = path.basename(filePath, '.js');
        
        // Basic JavaScript minification (simple approach)
        const minified = content
            // Remove single-line comments (but preserve URLs)
            .replace(/\/\/(?![^\n]*https?:)[^\n]*/g, '')
            // Remove multi-line comments
            .replace(/\/\*[\s\S]*?\*\//g, '')
            // Remove unnecessary whitespace
            .replace(/\s+/g, ' ')
            // Remove whitespace around operators and punctuation
            .replace(/\s*([{}();,=+\-*/<>!&|])\s*/g, '$1')
            // Remove trailing semicolons (optional)
            .replace(/;}/g, '}')
            .trim();
        
        const outputPath = path.join(this.outputDir, `${fileName}.min.js`);
        await fs.writeFile(outputPath, minified, 'utf8');
        
        const originalSize = Buffer.byteLength(content, 'utf8');
        const minifiedSize = Buffer.byteLength(minified, 'utf8');
        const savings = ((originalSize - minifiedSize) / originalSize * 100).toFixed(1);
        
        console.log(`  ✓ ${fileName}.js: ${originalSize} → ${minifiedSize} bytes (${savings}% reduction)`);
        
        return {
            original: filePath,
            minified: outputPath,
            content: minified,
            originalSize,
            minifiedSize,
            savings
        };
    }

    async createCombinedCSS(minifiedFiles) {
        // Order CSS files for optimal loading
        const orderedFiles = this.orderCSSFiles(minifiedFiles);
        
        const combinedContent = orderedFiles
            .map(file => `/* ${path.basename(file.original)} */\n${file.content}`)
            .join('\n\n');
        
        const combinedPath = path.join(this.outputDir, 'combined.min.css');
        await fs.writeFile(combinedPath, combinedContent, 'utf8');
        
        const totalSize = Buffer.byteLength(combinedContent, 'utf8');
        console.log(`  ✓ Combined CSS: ${totalSize} bytes`);
        
        return combinedPath;
    }

    async createCombinedJS(minifiedFiles) {
        // Order JS files for optimal loading
        const orderedFiles = this.orderJSFiles(minifiedFiles);
        
        const combinedContent = orderedFiles
            .map(file => `/* ${path.basename(file.original)} */\n${file.content}`)
            .join('\n\n');
        
        const combinedPath = path.join(this.outputDir, 'combined.min.js');
        await fs.writeFile(combinedPath, combinedContent, 'utf8');
        
        const totalSize = Buffer.byteLength(combinedContent, 'utf8');
        console.log(`  ✓ Combined JS: ${totalSize} bytes`);
        
        return combinedPath;
    }

    orderCSSFiles(files) {
        // Order CSS files for optimal cascade
        const order = ['main.css', 'components.css', 'responsive.css'];
        
        return files.sort((a, b) => {
            const aName = path.basename(a.original);
            const bName = path.basename(b.original);
            
            const aIndex = order.indexOf(aName);
            const bIndex = order.indexOf(bName);
            
            if (aIndex !== -1 && bIndex !== -1) {
                return aIndex - bIndex;
            } else if (aIndex !== -1) {
                return -1;
            } else if (bIndex !== -1) {
                return 1;
            } else {
                return aName.localeCompare(bName);
            }
        });
    }

    orderJSFiles(files) {
        // Order JS files for optimal dependency loading
        const order = ['image-optimizer.js', 'main.js', 'projects.js', 'publications.js', 'technical-artifacts.js', 'about.js', 'contact.js'];
        
        return files.sort((a, b) => {
            const aName = path.basename(a.original);
            const bName = path.basename(b.original);
            
            const aIndex = order.indexOf(aName);
            const bIndex = order.indexOf(bName);
            
            if (aIndex !== -1 && bIndex !== -1) {
                return aIndex - bIndex;
            } else if (aIndex !== -1) {
                return -1;
            } else if (bIndex !== -1) {
                return 1;
            } else {
                return aName.localeCompare(bName);
            }
        });
    }

    async generateManifest() {
        const manifest = {
            generated: new Date().toISOString(),
            version: '1.0.0',
            files: {
                css: {
                    combined: 'assets/dist/combined.min.css',
                    individual: []
                },
                js: {
                    combined: 'assets/dist/combined.min.js',
                    individual: []
                }
            },
            optimization: {
                minification: true,
                compression: true,
                caching: true
            }
        };
        
        // Add individual files
        const cssFiles = await this.findFiles(this.outputDir, '.min.css');
        const jsFiles = await this.findFiles(this.outputDir, '.min.js');
        
        manifest.files.css.individual = cssFiles
            .filter(file => !file.includes('combined'))
            .map(file => path.relative('.', file));
            
        manifest.files.js.individual = jsFiles
            .filter(file => !file.includes('combined'))
            .map(file => path.relative('.', file));
        
        const manifestPath = path.join(this.outputDir, 'build-manifest.json');
        await fs.writeFile(manifestPath, JSON.stringify(manifest, null, 2), 'utf8');
        
        console.log('✓ Build manifest created');
    }
}

// CLI execution
if (require.main === module) {
    const minifier = new MinificationTool();
    minifier.init().catch(console.error);
}

module.exports = MinificationTool;